pkg/wal/wal.go

```diff
@@ -316,17 +342,28 @@ func (w *WAL) Remove(path string) error {
 }
 
 func (w *WAL) Append(ctx context.Context, buf []byte) error {
-	err := w.tryAppend(ctx, buf)
-	if errors.Is(err, ErrSegmentClosed) {
-		return w.tryAppend(ctx, buf)
+	atomic.AddInt64(&w.inflightWriteBytes, int64(len(buf)))
+	defer atomic.AddInt64(&w.inflightWriteBytes, -int64(len(buf)))
+
+	n, err := w.tryAppend(ctx, buf)
+	if errors.Is(err, ErrMaxSegmentSizeExceeded) {
+		w.rotateSegmentIfNecessary()
+		n, err = w.tryAppend(ctx, buf)
+		atomic.AddInt64(&w.segmentSize, int64(n))
+		return err
```

Comment:
if `tryAppend` returns no error, it will overwrite the returned error. 

```suggestion
		n, appendErr := w.tryAppend(ctx, buf)
		atomic.AddInt64(&w.segmentSize, int64(n))
		return errors.Join(err, appendErr)
```

Reply:
This intended because we are handling the ErrMaxSegmentSizeExceeded error here.  When that error is detected, we rotate the segment.  We don't want to return the ErrMaxSegmentSizeExceeded error.